### ISSUE TITLE
Rust analyzer false positive error

### ISSUE THREAD

**Issue by benluelo** [rust-lang/rust-analyzer#18133]
```
$ rust-analyzer --version
rust-analyzer 1.83.0-nightly (c52c23b6 2024-09-16)

$ rustc -V --verbose
rustc 1.83.0-nightly (c52c23b6f 2024-09-16)
binary: rustc
commit-hash: c52c23b6f44cd19718721a5e3b2eeb169e9c96ff
commit-date: 2024-09-16
host: aarch64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0
```

```rust
pub enum Enum<T: Foo> {
    String(String),
    ImplsBar(T::UsedInEnum),
}

impl<T: Foo> Enum<T> {
    pub fn f(&self) {
        match *self {
            Enum::String(ref string) => dyn_bar(&string),
            Enum::ImplsBar(ref a1) => dyn_bar(&a1),
        };
    }
}

pub trait Foo {
    type UsedInEnum: Bar;
    type Unused1: Quux;
    type Unused2: Quux;
}

pub trait Quux {
    type Unused3: Bar;
}

pub trait Bar {}

impl Bar for String {}

impl<T: Bar> Bar for &'_ T {}

pub fn dyn_bar(_: &dyn Bar) {}
```

`cargo check` works fine, however `rust-analyzer` errors:

```
$ rust-analyzer diagnostics .
processing crate: playground, module: /redacted/src/main.rs
Error RustcHardError("E0308") from LineCol { line: 10, col: 48 } to LineCol { line: 10, col: 55 }: expected &dyn Bar, found &&String
Error RustcHardError("E0308") from LineCol { line: 11, col: 46 } to LineCol { line: 11, col: 49 }: expected &dyn Bar, found &&<T as Foo>::UsedInEnum

diagnostic scan complete

Error: diagnostic error detected

Stack backtrace:
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2: std::backtrace::Backtrace::create
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/std/src/backtrace.rs:331:13
   3: <anyhow::Error>::msg::<&str>
   4: std::sys::backtrace::__rust_begin_short_backtrace::<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>
   5: <<std::thread::Builder>::spawn_unchecked_<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
   6: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/alloc/src/boxed.rs:2453:9
   7: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/alloc/src/boxed.rs:2453:9
   8: std::sys::pal::unix::thread::Thread::new::thread_start
             at /rustc/c52c23b6f44cd19718721a5e3b2eeb169e9c96ff/library/std/src/sys/pal/unix/thread.rs:105:17
   9: start_thread
  10: thread_start
```

Notably both `String` and `T::UsedInEnum` error. However! If any of the `Unused*` associated types are commented out, it compiles.

**Comment by ShoyuVanilla** [comment]
I've tried a bit but this seems to be a chalk issue 😢

**Comment by benluelo** [comment]
I've bisected this to:

rust-analyzer 1.82.0-nightly (0d634185 2024-08-29) -> works
rust-analyzer 1.82.0-nightly (a7399ba6 2024-08-31) -> doesn't work

**Comment by ShoyuVanilla** [comment]
> I've bisected this to:
> 
> rust-analyzer 1.82.0-nightly (0d634185 2024-08-29) -> works rust-analyzer 1.82.0-nightly (a7399ba6 2024-08-31) -> doesn't work

Oh, that's cool. This might not be a chalk issue then. I'll look into this more. Thanks 😄

**Comment by ShoyuVanilla** [comment]
Well, I've tried with the following nightly versions but these all emitted the same false positive diagnostics
```
rust-analyzer version: 0.4.2077-standalone (fa0032624 2024-08-17)
rust-analyzer version: 0.4.2081-standalone (3723e5910 2024-08-20)
rust-analyzer version: 0.4.2090-standalone (8db40df2a 2024-08-27)
rust-analyzer version: 0.4.2091-standalone (f454ea877 2024-08-28)
rust-analyzer version: 0.4.2092-standalone (0ae42bd42 2024-08-29)
rust-analyzer version: 0.4.2093-standalone (9fd705195 2024-08-30)
```
And the short hashes you've mentioned (`0d634185` and `a7399ba6`) don't exist in the rust-analyzer's commit graph.

With those observations and looking into r-a codes I doubt if this is a new issue that has recently emerged.

How did you get those versions, `0d634185` and `a7399ba6`?

**Comment by ShoyuVanilla** [comment]
My bisection result is a bit different

rust-analyzer version: 0.4.2075-standalone (07659783f 2024-08-16) -> Makes those false positives
rust-analyzer version: 0.4.2074-standalone (91aa3f46b 2024-08-15) -> No false positives

**Comment by benluelo** [comment]
I use https://github.com/oxalica/rust-overlay - I'm unable to get those same commits as you seemingly:

rust-analyzer 1.82.0-nightly (2c93fabd 2024-08-15)
rust-analyzer 1.82.0-nightly (506052d4 2024-08-16)

(both of these work for me also)

**Comment by ShoyuVanilla** [comment]
Thanks!
I've found the most suspicious single commit.
Will work on this tommorrow but not sure if this could be fixed soon 😅

**Comment by ShoyuVanilla** [comment]
It seems that nix rust overlay prints versions from its own commit graphs btw

**Comment by ShoyuVanilla** [comment]
Bisected to 8a4261aca83bdf83db02808d7dbb030bb0e4fcc2

**Comment by ShoyuVanilla** [comment]
As @dfireBird is [working on lifetime elision](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/HIR.20lowering.20for.20lifetimes/near/468074719), this seems to be resolved once his work is done. If this isn't fixed then, I'll look into this more.

**Comment by jplatte** [comment]
Seems to be working now, with the new trait solver.
